### ISSUE TITLE
You have declared one of kalvidassign_grade_item_update and kalvidassign_update_grades but not both. This will cause broken behaviour.

### ISSUE THREAD

**Issue by rlorenzo** [kaltura/moodle_plugin#156]
We are running 2017062531 of mod_kalvidassign and when you have a Kaltura Video Assignment and you enroll someone into the course you get this error:

```
You have declared one of kalvidassign_grade_item_update and kalvidassign_update_grades but not both. This will cause broken behaviour.
line 1403 of /lib/gradelib.php: call to debugging()
line 1369 of /lib/gradelib.php: call to grade_update_mod_grades()
line 1123 of /lib/gradelib.php: call to grade_grab_course_grades()
line 1368 of /lib/enrollib.php: call to grade_recover_history_grades()
line 675 of /enrol/database/lib.php: call to enrol_plugin->enrol_user()
```

It seems only kalvidassign_grade_item_update() is implemented. We are running Moodle 3.1.7.

I checked the lib.php for the MOODLE_33_STABLE branch and it is missing an implementation of kalvidassign_update_grades() as well.

**Comment by muli** [comment]
Hi @rlorenzo,

I tried to reproduce this issue according to your description and I wasn't able to.
Can you provide more specific steps?

Thank you.

**Comment by rlorenzo** [comment]
@muli To reproduce this you can reerolling someone who you remove their enrollment before. But do you really need steps to reproduce when the debugging message is quite clear that the code is doing something that is not expected in the Moodle API?

See https://github.com/moodle/moodle/blob/daf0b4f08be69c34550f9df47cef43b17473c372/lib/gradelib.php#L1400

To support grading both kalvidassign_grade_item_update() and kalvidassign_update_grades() need to be implemented.

**Comment by muli** [comment]
@rlorenzo, thanks.

The reason why I asked for steps to reproduce is simply in order to:
1. Test that my attempted fix works.
2. Provide our QA team instructions for testing this fix.

**Comment by rlorenzo** [comment]
@muli Steps to reproduce (might need to turn on Moodle debugging mode)

1) Make sure you have a course with a Kaltura Video Assignment
2) Go to a course and the participants page
3) Enroll someone into the course as a student. In the manual enrollment modal click on "Show more" and make sure "Recover user's old grades if possible" is checked.
4) Check the PHP error log and you should see an entry called as follows:

[29-Nov-2018 10:41:38 America/Los_Angeles] Debugging: You have declared one of kalvidassign_grade_item_update and kalvidassign_update_grades but not both. This will cause broken behaviour. in 
* line 1405 of /lib/gradelib.php: call to debugging()
* line 1371 of /lib/gradelib.php: call to grade_update_mod_grades()
* line 1125 of /lib/gradelib.php: call to grade_grab_course_grades()
* line 1915 of /lib/enrollib.php: call to grade_recover_history_grades()
* line 143 of /enrol/manual/ajax.php: call to enrol_plugin->enrol_user()

**Comment by muli** [comment]
Thank you very much! I was able to reproduce and a fix will be issued soon.

**Comment by muli** [comment]
Fixed in v4.1.4 for Moodle 3.4, 3.5 and 3.6.
